dbms/src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -2057,44 +2062,17 @@ void MergeTreeData::freezePartition(const ASTPtr & partition_ast, const String &
     else
         LOG_DEBUG(log, "Freezing parts with partition ID " + partition_id);
 
-    String clickhouse_path = Poco::Path(context.getPath()).makeAbsolute().toString();
-    String shadow_path = clickhouse_path + "shadow/";
-    Poco::File(shadow_path).createDirectories();
-    String backup_path = shadow_path
-        + (!with_name.empty()
-            ? escapeForFileName(with_name)
-            : toString(Increment(shadow_path + "increment.txt").get(true)))
-        + "/";
-
-    LOG_DEBUG(log, "Snapshot will be placed at " + backup_path);
 
-    /// Acquire a snapshot of active data parts to prevent removing while doing backup.
-    const auto data_parts = getDataParts();
-
-    size_t parts_processed = 0;
-    for (const auto & part : data_parts)
-    {
-        if (prefix)
+    freezePartitionsByMatcher(
+        [&prefix, &partition_id](DataPartPtr part)
```

Comment:
Why don't pass shared_ptr by reference to avoid atomic ops?

Reply:
(it would not affect performance in this case, more a matter of style,
because copying of shared_ptr raises unnecessarily concerns)